Building Dynamic Framework on iOS without some modules

##### System information (version)

- OpenCV => HEAD (ece3fac)
- Operating System / Platform => macOS 10.12.3 (16D32)
- Compiler => Apple LLVM version 8.0.0 (clang-800.0.42.1)

##### Basic description

I can not build a dynamic framework for iOS while excluding some modules.
```
Ld /tmp/opencv3/ios/build/build-iphoneos/modules/imgcodecs/OpenCV.build/Release-iphoneos/opencv_imgcodecs.build/Objects-normal/arm64/opencv2 normal arm64
    cd /tmp/opencv3/opencv
    export IPHONEOS_DEPLOYMENT_TARGET=8.0
    export PATH="/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/usr/bin:/Applications/Xcode.app/Contents/Developer/usr/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Java/JavaVirtualMachines/jdk1.8.0_121.jdk/Contents/Home"
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -arch arm64 -dynamiclib -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS10.2.sdk -L/tmp/opencv3/ios/build/build-iphoneos/lib/Release -F/tmp/opencv3/ios/build/build-iphoneos/lib/Release -filelist /tmp/opencv3/ios/build/build-iphoneos/modules/imgcodecs/OpenCV.build/Release-iphoneos/opencv_imgcodecs.build/Objects-normal/arm64/opencv2.LinkFileList -install_name @rpath/opencv2.framework/opencv2 -miphoneos-version-min=8.0 -dead_strip -fembed-bitcode-marker -fobjc-link-runtime -rpath @executable_path/Frameworks -rpath @loader_path/Frameworks /tmp/opencv3/ios/build/build-iphoneos/lib/Release/opencv2.framework/opencv2 /tmp/opencv3/ios/build/build-iphoneos/3rdparty/lib/Release/libzlib.a /tmp/opencv3/ios/build/build-iphoneos/3rdparty/lib/Release/liblibjpeg.a /tmp/opencv3/ios/build/build-iphoneos/3rdparty/lib/Release/liblibpng.a -framework Accelerate -framework CoreGraphics -framework QuartzCore -framework AssetsLibrary /tmp/opencv3/ios/build/build-iphoneos/lib/Release/opencv2.framework/opencv2 /tmp/opencv3/ios/build/build-iphoneos/3rdparty/lib/Release/libzlib.a -single_module -compatibility_version 3.2.0 -current_version 3.2.0 -Xlinker -dependency_info -Xlinker /tmp/opencv3/ios/build/build-iphoneos/modules/imgcodecs/OpenCV.build/Release-iphoneos/opencv_imgcodecs.build/Objects-normal/arm64/opencv2_dependency_info.dat -o /tmp/opencv3/ios/build/build-iphoneos/modules/imgcodecs/OpenCV.build/Release-iphoneos/opencv_imgcodecs.build/Objects-normal/arm64/opencv2
Undefined symbols for architecture arm64:
  "_cvGetMat", referenced from:
      _cvConvertImage in utils.o
  "_cvError", referenced from:
      _cvConvertImage in utils.o
  "cv::transpose(cv::_InputArray const&, cv::_OutputArray const&)", referenced from:
      cv::ApplyExifOrientation(cv::String const&, cv::Mat&) in loadsave.o
  "cv::cvarrToMat(void const*, bool, bool, int, cv::AutoBuffer<double, 136ul>*)", referenced from:
      cv::imread_(cv::String const&, int, int, cv::Mat*) in loadsave.o
      cv::imdecode_(cv::Mat const&, int, int, cv::Mat*) in loadsave.o
      _cvSaveImage in loadsave.o
      _cvEncodeImage in loadsave.o
  "cv::_InputArray::kind() const", referenced from:
      cv::imwrite(cv::String const&, cv::_InputArray const&, std::__1::vector<int, std::__1::allocator<int> > const&) in loadsave.o
      cv::imdecode(cv::_InputArray const&, int) in loadsave.o
      cv::imdecode(cv::_InputArray const&, int, cv::Mat*) in loadsave.o
      cv::imencode(cv::String const&, cv::_InputArray const&, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >&, std::__1::vector<int, std::__1::allocator<int> > const&) in loadsave.o
  "_cvCreateImage", referenced from:
      cv::imread_(cv::String const&, int, int, cv::Mat*) in loadsave.o
      cv::imdecode_(cv::Mat const&, int, int, cv::Mat*) in loadsave.o
  "_cvGetErrStatus", referenced from:
      _cvConvertImage in utils.o
  "cv::_InputArray::getMat_(int) const", referenced from:
      cv::imwrite(cv::String const&, cv::_InputArray const&, std::__1::vector<int, std::__1::allocator<int> > const&) in loadsave.o
      cv::imdecode(cv::_InputArray const&, int) in loadsave.o
      cv::imdecode(cv::_InputArray const&, int, cv::Mat*) in loadsave.o
      cv::imencode(cv::String const&, cv::_InputArray const&, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >&, std::__1::vector<int, std::__1::allocator<int> > const&) in loadsave.o
  "_cvCreateMat", referenced from:
      cv::imread_(cv::String const&, int, int, cv::Mat*) in loadsave.o
      cv::imdecode_(cv::Mat const&, int, int, cv::Mat*) in loadsave.o
      _cvEncodeImage in loadsave.o
      _cvConvertImage in utils.o
  "_cvCopy", referenced from:
      _cvConvertImage in utils.o
  "_cvReleaseImage", referenced from:
      cv::imread_(cv::String const&, int, int, cv::Mat*) in loadsave.o
      cv::imdecode_(cv::Mat const&, int, int, cv::Mat*) in loadsave.o
  "_cvReleaseMat", referenced from:
      cv::imread_(cv::String const&, int, int, cv::Mat*) in loadsave.o
      cv::imdecode_(cv::Mat const&, int, int, cv::Mat*) in loadsave.o
      _cvConvertImage in utils.o
  "cv::String::deallocate()", referenced from:
      cv::RBaseStream::open(cv::Mat const&) in bitstrm.o
      cv::BaseImageDecoder::setSource(cv::String const&) in grfmt_base.o
      cv::BaseImageDecoder::setSource(cv::Mat const&) in grfmt_base.o
      cv::BaseImageEncoder::setDestination(cv::String const&) in grfmt_base.o
      cv::BaseImageEncoder::setDestination(std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >&) in grfmt_base.o
      cv::BaseImageEncoder::throwOnEror() const in grfmt_base.o
      cv::BaseImageEncoder::~BaseImageEncoder() in grfmt_base.o
      ...
  "cv::Mat::copySize(cv::Mat const&)", referenced from:
      cv::Mat::operator=(cv::Mat const&) in grfmt_base.o
      std::__1::vector<cv::Mat, std::__1::allocator<cv::Mat> >::vector(unsigned long, cv::Mat const&) in grfmt_hdr.o
      cv::imreadmulti(cv::String const&, std::__1::vector<cv::Mat, std::__1::allocator<cv::Mat> >&, int) in loadsave.o
      cv::imwrite(cv::String const&, cv::_InputArray const&, std::__1::vector<int, std::__1::allocator<int> > const&) in loadsave.o
      cv::imdecode(cv::_InputArray const&, int) in loadsave.o
      cv::imdecode(cv::_InputArray const&, int, cv::Mat*) in loadsave.o
      cv::imencode(cv::String const&, cv::_InputArray const&, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >&, std::__1::vector<int, std::__1::allocator<int> > const&) in loadsave.o
      ...
  "_OBJC_CLASS_$_UIImage", referenced from:
      objc-class-ref in ios_conversions.o
  "_cvConvertScale", referenced from:
      _cvConvertImage in utils.o
  "cv::error(int, cv::String const&, char const*, char const*, int)", referenced from:
      cv::RBaseStream::open(cv::Mat const&) in bitstrm.o
      cv::BaseImageEncoder::throwOnEror() const in grfmt_base.o
      cv::HdrEncoder::write(cv::Mat const&, std::__1::vector<int, std::__1::allocator<int> > const&) in grfmt_hdr.o
      cv::PngDecoder::readDataFromBuf(void*, unsigned char*, unsigned long) in grfmt_png.o
      cv::PngEncoder::writeDataToBuf(void*, unsigned char*, unsigned long) in grfmt_png.o
      cv::imwrite_(cv::String const&, cv::Mat const&, std::__1::vector<int, std::__1::allocator<int> > const&, bool) in loadsave.o
      cv::imdecode_(cv::Mat const&, int, int, cv::Mat*) in loadsave.o
      ...
  "cv::merge(cv::_InputArray const&, cv::_OutputArray const&)", referenced from:
      cv::HdrEncoder::write(cv::Mat const&, std::__1::vector<int, std::__1::allocator<int> > const&) in grfmt_hdr.o
  "cv::Mat::create(int, int const*, int)", referenced from:
      cv::HdrDecoder::readData(cv::Mat&) in grfmt_hdr.o
      cv::imread_(cv::String const&, int, int, cv::Mat*) in loadsave.o
      cv::imreadmulti(cv::String const&, std::__1::vector<cv::Mat, std::__1::allocator<cv::Mat> >&, int) in loadsave.o
      cv::imdecode_(cv::Mat const&, int, int, cv::Mat*) in loadsave.o
  "cv::fastFree(void*)", referenced from:
      cv::BaseImageDecoder::~BaseImageDecoder() in grfmt_base.o
      cv::HdrDecoder::readData(cv::Mat&) in grfmt_hdr.o
      cv::HdrEncoder::write(cv::Mat const&, std::__1::vector<int, std::__1::allocator<int> > const&) in grfmt_hdr.o
      std::__1::__vector_base<cv::Mat, std::__1::allocator<cv::Mat> >::~__vector_base() in grfmt_hdr.o
      cv::imread(cv::String const&, int) in loadsave.o
      cv::imread_(cv::String const&, int, int, cv::Mat*) in loadsave.o
      cv::imreadmulti(cv::String const&, std::__1::vector<cv::Mat, std::__1::allocator<cv::Mat> >&, int) in loadsave.o
      ...
  "cv::Mat::convertTo(cv::_OutputArray const&, int, double, double) const", referenced from:
      cv::HdrDecoder::readData(cv::Mat&) in grfmt_hdr.o
      cv::HdrEncoder::write(cv::Mat const&, std::__1::vector<int, std::__1::allocator<int> > const&) in grfmt_hdr.o
      cv::imwrite_(cv::String const&, cv::Mat const&, std::__1::vector<int, std::__1::allocator<int> > const&, bool) in loadsave.o
      cv::imencode(cv::String const&, cv::_InputArray const&, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >&, std::__1::vector<int, std::__1::allocator<int> > const&) in loadsave.o
  "cv::Mat::deallocate()", referenced from:
      cv::BaseImageDecoder::setSource(cv::String const&) in grfmt_base.o
      cv::Mat::operator=(cv::Mat const&) in grfmt_base.o
      cv::BaseImageDecoder::~BaseImageDecoder() in grfmt_base.o
      cv::HdrDecoder::readData(cv::Mat&) in grfmt_hdr.o
      cv::HdrEncoder::write(cv::Mat const&, std::__1::vector<int, std::__1::allocator<int> > const&) in grfmt_hdr.o
      std::__1::__vector_base<cv::Mat, std::__1::allocator<cv::Mat> >::~__vector_base() in grfmt_hdr.o
      cv::imread(cv::String const&, int) in loadsave.o
      ...
  "_cvFlip", referenced from:
      _cvConvertImage in utils.o
  "cv::Mat::copyTo(cv::_OutputArray const&) const", referenced from:
      cv::HdrEncoder::write(cv::Mat const&, std::__1::vector<int, std::__1::allocator<int> > const&) in grfmt_hdr.o
  "cv::tempfile(char const*)", referenced from:
      cv::imdecode_(cv::Mat const&, int, int, cv::Mat*) in loadsave.o
      cv::imencode(cv::String const&, cv::_InputArray const&, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >&, std::__1::vector<int, std::__1::allocator<int> > const&) in loadsave.o
  "cv::flip(cv::_InputArray const&, cv::_OutputArray const&, int)", referenced from:
      cv::ApplyExifOrientation(cv::String const&, cv::Mat&) in loadsave.o
      cv::imwrite_(cv::String const&, cv::Mat const&, std::__1::vector<int, std::__1::allocator<int> > const&, bool) in loadsave.o
      _cvEncodeImage in loadsave.o
  "cv::String::allocate(unsigned long)", referenced from:
      cv::RBaseStream::open(cv::Mat const&) in bitstrm.o
      cv::BaseImageEncoder::throwOnEror() const in grfmt_base.o
      cv::BmpDecoder::BmpDecoder() in grfmt_bmp.o
      cv::BmpEncoder::BmpEncoder() in grfmt_bmp.o
      cv::HdrDecoder::HdrDecoder() in grfmt_hdr.o
      cv::HdrEncoder::HdrEncoder() in grfmt_hdr.o
      cv::HdrEncoder::write(cv::Mat const&, std::__1::vector<int, std::__1::allocator<int> > const&) in grfmt_hdr.o
      ...
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

** BUILD FAILED **


The following build commands failed:
	Ld /tmp/opencv3/ios/build/build-iphoneos/modules/imgcodecs/OpenCV.build/Release-iphoneos/opencv_imgcodecs.build/Objects-normal/arm64/opencv2 normal arm64
(1 failure)
============================================================
ERROR: Command '['xcodebuild', 'IPHONEOS_DEPLOYMENT_TARGET=8.0', 'ONLY_ACTIVE_ARCH=NO', '-arch', 'armv7', '-arch', 'armv7s', '-arch', 'arm64', '-sdk', 'iphoneos', '-configuration', 'Release', '-parallelizeTargets', '-jobs', '4', '-target', 'ALL_BUILD', '-target', 'ALL_BUILD', 'build']' returned non-zero exit status 65
============================================================
Traceback (most recent call last):
  File "opencv/platforms/ios/build_framework.py", line 112, in build
    self._build(outdir)
  File "opencv/platforms/ios/build_framework.py", line 104, in _build
    self.buildOne(t[0], t[1], mainBD, cmake_flags)
  File "opencv/platforms/ios/build_framework.py", line 204, in buildOne
    execute(buildcmd + ["-target", "ALL_BUILD", "build"], cwd = builddir)
  File "opencv/platforms/ios/build_framework.py", line 36, in execute
    retcode = check_call(cmd, cwd = cwd)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
CalledProcessError: Command '['xcodebuild', 'IPHONEOS_DEPLOYMENT_TARGET=8.0', 'ONLY_ACTIVE_ARCH=NO', '-arch', 'armv7', '-arch', 'armv7s', '-arch', 'arm64', '-sdk', 'iphoneos', '-configuration', 'Release', '-parallelizeTargets', '-jobs', '4', '-target', 'ALL_BUILD', '-target', 'ALL_BUILD', 'build']' returned non-zero exit status 65

```

##### Detailed description

Building OpenCV as dynamic framework w/o excluding any modules at all works perfectly.

I deep dived into #7358 and #8009 but could not find a reason for this behaviour.

Maybe @alalek can help out here.

Thanks in advance!

Cheers!

##### Steps to reproduce

```bash
git clone git@github.com:opencv/opencv.git
git checkout ece3fac # which is as of now the current HEAD
```

### does not work at all

```bash
python opencv/platforms/ios/build_framework.py ios \
    --dynamic \
    --without video --without videoio --without videostab \
    --without features2d --without objdetect --without flann --without ml \
    --without cudaarithm --without cudabgsegm --without cudacodec --without cudafeatures2d  \
    --without cudafilters --without cudaimgproc --without cudalegacy --without cudaobjdetect  \
    --without cudaoptflow --without cudastereo --without cudawarping --without cudev \
    --without highgui --without viz \
    --without superres --without photo
```
</details>

### works

Leaving out the dynamic parameter leads to a successful build.

```bash
python opencv/platforms/ios/build_framework.py ios \
    --without video --without videoio --without videostab \
    --without features2d --without objdetect --without flann --without ml \
    --without cudaarithm --without cudabgsegm --without cudacodec --without cudafeatures2d  \
    --without cudafilters --without cudaimgproc --without cudalegacy --without cudaobjdetect  \
    --without cudaoptflow --without cudastereo --without cudawarping --without cudev \
    --without highgui --without viz \
    --without superres --without photo
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -519,7 +519,9 @@ endif(WIN32 AND NOT MINGW)
 #       CHECK FOR SYSTEM LIBRARIES, OPTIONS, ETC..
 # ----------------------------------------------------------------------------
 if(UNIX)
-  find_package(PkgConfig QUIET)
+  if(NOT APPLE_FRAMEWORK)
+    find_package(PkgConfig QUIET)
+  endif()
   include(CheckFunctionExists)
   include(CheckIncludeFile)
 

--- a/modules/imgcodecs/CMakeLists.txt
+++ b/modules/imgcodecs/CMakeLists.txt
@@ -92,6 +92,9 @@ if(IOS)
   list(APPEND imgcodecs_srcs ${CMAKE_CURRENT_LIST_DIR}/src/ios_conversions.mm)
   list(APPEND IMGCODECS_LIBRARIES "-framework Accelerate" "-framework CoreGraphics" "-framework QuartzCore" "-framework AssetsLibrary")
 endif()
+if(APPLE_FRAMEWORK)
+  list(APPEND IMGCODECS_LIBRARIES "-framework UIKit")
+endif()
 
 if(UNIX)
   #these variables are set by CHECK_MODULE macro

--- a/modules/videoio/CMakeLists.txt
+++ b/modules/videoio/CMakeLists.txt
@@ -210,7 +210,7 @@ if(IOS)
        ${CMAKE_CURRENT_LIST_DIR}/src/cap_ios_video_camera.mm)
 
   list(APPEND VIDEOIO_LIBRARIES "-framework Accelerate" "-framework AVFoundation" "-framework CoreGraphics" "-framework CoreImage" "-framework CoreMedia" "-framework CoreVideo" "-framework QuartzCore" "-framework AssetsLibrary")
-  if(APPLE_FRAMEWORK AND BUILD_SHARED_LIBS)
+  if(APPLE_FRAMEWORK)
     list(APPEND VIDEOIO_LIBRARIES "-framework UIKit")
   endif()
 endif()

--- a/platforms/ios/build_framework.py
+++ b/platforms/ios/build_framework.py
@@ -122,37 +122,32 @@ class Builder:
 
     def getCMakeArgs(self, arch, target):
 
-        if self.dynamic:
-            args = [
-                "cmake",
-                "-GXcode",
-                "-DAPPLE_FRAMEWORK=ON",
-                "-DCMAKE_INSTALL_PREFIX=install",
-                "-DCMAKE_BUILD_TYPE=Release",
-                "-DBUILD_SHARED_LIBS=ON",
-                "-DCMAKE_MACOSX_BUNDLE=ON",
-                "-DCMAKE_XCODE_ATTRIBUTE_CODE_SIGNING_REQUIRED=NO",
-            ]
-        else:
-            args = [
-                "cmake",
-                "-GXcode",
-                "-DAPPLE_FRAMEWORK=ON",
-                "-DCMAKE_INSTALL_PREFIX=install",
-                "-DCMAKE_BUILD_TYPE=Release",
-            ]
+        args = [
+            "cmake",
+            "-GXcode",
+            "-DAPPLE_FRAMEWORK=ON",
+            "-DCMAKE_INSTALL_PREFIX=install",
+            "-DCMAKE_BUILD_TYPE=Release",
+        ] + ([
+            "-DBUILD_SHARED_LIBS=ON",
+            "-DCMAKE_MACOSX_BUNDLE=ON",
+            "-DCMAKE_XCODE_ATTRIBUTE_CODE_SIGNING_REQUIRED=NO",
+        ] if self.dynamic else [])
 
         if len(self.exclude) > 0:
             args += ["-DBUILD_opencv_world=OFF"]
-            args += ("-DBUILD_opencv_%s=OFF" % m for m in self.exclude)
+            args += ["-DBUILD_opencv_%s=OFF" % m for m in self.exclude]
 
         return args
 
     def getBuildCommand(self, archs, target):
 
+        buildcmd = [
+            "xcodebuild",
+        ]
+
         if self.dynamic:
-            buildcmd = [
-                "xcodebuild",
+            buildcmd += [
                 "IPHONEOS_DEPLOYMENT_TARGET=8.0",
                 "ONLY_ACTIVE_ARCH=NO",
             ]
@@ -160,25 +155,19 @@ class Builder:
             for arch in archs:
                 buildcmd.append("-arch")
                 buildcmd.append(arch.lower())
-
+        else:
+            arch = ";".join(archs)
             buildcmd += [
+                "IPHONEOS_DEPLOYMENT_TARGET=6.0",
+                "ARCHS=%s" % arch,
+            ]
+
+        buildcmd += [
                 "-sdk", target.lower(),
                 "-configuration", "Release",
                 "-parallelizeTargets",
                 "-jobs", "4",
-                "-target","ALL_BUILD",
-            ]
-        else:
-            arch = ";".join(archs)
-            buildcmd = [
-                "xcodebuild",
-                "IPHONEOS_DEPLOYMENT_TARGET=6.0",
-                "ARCHS=%s" % arch,
-                "-sdk", target.lower(),
-                "-configuration", "Release",
-                "-parallelizeTargets",
-                "-jobs", "4"
-            ]
+            ] + (["-target","ALL_BUILD"] if self.dynamic else [])
 
         return buildcmd
 

--- a/platforms/ios/build_framework.py
+++ b/platforms/ios/build_framework.py
@@ -135,7 +135,7 @@ class Builder:
         ] if self.dynamic else [])
 
         if len(self.exclude) > 0:
-            args += ["-DBUILD_opencv_world=OFF"]
+            args += ["-DBUILD_opencv_world=OFF"] if not self.dynamic else []
             args += ["-DBUILD_opencv_%s=OFF" % m for m in self.exclude]
 
         return args

